### DISCUSSION TITLE
Multiple dynamic segments matching single path

### DISCUSSION THREAD

**Discussion by piotrkulpinski** [remix-run/remix#4794] [Q&A]
Hi, I'm building an app with many dynamic paths involved. 

I searched the docs but can't find anything that would allow me to achieve the following. 

--- 

My Route structure looks like this:
```bash
$postType
- $collectionType
- - $id.tsx
- - index.tsx
- $id.tsx
- index.tsx
```

So it's basically a structure of post and collection types that gives me possible paths like this:
- `/books` - renders `$postType/index.tsx`
- `/books/the-shining` - renders `$postType/$id.tsx` dynamically
- `/books/authors` - renders `$postType/$collectionType/index.tsx`
- `/books/authors/steven-king` - renders `$postType/$collectionType/$id.tsx` dynamically

Unfortunately, out of the box, it takes only the first dynamic path it finds, so `/books/the-shining` would point to the `$postType/$collectionType/index.tsx` file. 

Is there any way to tell Remix from the loader, that the collection type with that slug was not found and to go to another dynamic path?

Thanks a lot for any help!

**Comment by jameslai** [comment]
@piotrkulpinski Did you ever figure out a strategy to this? I too am facing a very similar issue. As my use-case involves multiple levels of dynamic segments, it seems to me Remix's routing strategy fails almost immediately and I'm left with a single splat file and loader, hand-rolling all routing myself. I'm hoping you might have found a better answer than that!

EDIT: Alright, I figured out something that is at least viable, using custom routes: https://remix.run/docs/en/main/file-conventions/remix-config#routes

Basically define everything within the remix.config.js file, which does seem to completely support multiple dynamic segments.

**Reply by piotrkulpinski** [reply]
Unfortunately not. I’m handling all the complex routing in a single route file. I’m giving one content type a priority and if it’s not found, I look further.

Interesting find about the route config. Could you share your config once you figure it out?

**Reply by jameslai** [reply]
You bet - it looks a little something like this:

```js
const {defineRoutes} = require("@remix-run/dev/dist/config/routes");

const routes = async (defineRoutes) => {
  return defineRoutes(route => {
    route("/:foo/:bar", "routes/foo.tsx")
    route("/:foo/:bar", "routes/bar.tsx")
    route("/:foo/:bar/*", "routes/baz.tsx")
  })
}

module.exports = {
   // Normal config
   routes
}

```

So the upsides:
 - The normal exported `loader` constant works! Great, no need for a master loader anymore
 - The normal routing within the `routes` directory still works - if the file isn't already included in the routes config displayed above, which is great! I was worried the behavior might be the file would be rendered in two locations
 
Downsides I've noticed:
- You can only register a file to a _single_ route. So if you wanted `/:foo/:bar` and `/:foo/:bar/*` to render the same thing (not uncommon imho), it won't let you specify the same file. I ended up making a "dummy" file that simply re-exported the `loader` and default export of the intended route.
- The dev mode of Remix seems to not watch remix.config.js, so changing routes here will require restarting the service
- This does not appear to work with route nesting
